proposal/proposal.tex

```diff
@@ -225,50 +211,77 @@ \subsection{Hygiene}
 Point \ref{responsibility}. above touches on an important topic: PPXs generate
 the final representation of the program and so have an impact on how the
 compiler and the editor tool $merlin$ analyze the program. That's why we define
-clear hygienic rules\cite{hygiene} and try to create a culture in which PPX
-authors are aware of the possible consequences.
+clear hygiene rules\cite{hygiene} and try to create a culture in which PPX
+authors are aware of the possible consequences. Some of these hygiene rules are
+enforced technically by $ppxlib$, and others are currently only guidelines that
+PPX authors should follow. The PPX mechanism within the compiler itself does not
+enforce any technical notion of hygiene as in some macro
+systems\cite{scheme_hygiene}.
 
 \subsubsection*{Error handling}
 
-It's important to handle PPX errors at the level of the AST abstraction.
-Concretely, errors need to be embedded as nodes into the AST. A raising PPX
-means that the compiler and $merlin$ don't receive an expanded AST.
-%and won't be able to reason about the whole program.
+Errors in OCaml are most often handled by raising exceptions. Naive error
+handling within the $ppxlib$ driver would abort transformation of an entire
+program the first time an exception was raised. This means tools like $merlin$
+learn nothing about the shape of the transformed program, and debugging the
+problem becomes only more difficult.
+
+Our first approach to addressing this was to provide tools for expressing errors
+as AST nodes, and encouraging PPX authors to use these instead of exceptions.
+This practice was slow to creep into the ecosystem, so we made the behavior
+automatic by handling exceptions inside $ppxlib$. Now an exception raised by a
+context-free transformation is caught, and its error message is stored as an AST
+node so that the rest of the program can be transformed.
 
 \subsubsection*{Location invariants}
 
 The \emph{Location} of a AST node reveals which part of the source code the node
-corresponds to.
+corresponds to. We've run into some problems with transformed code, wherein
+tools like $merlin$ cannot figure out which piece of transformed or duplicated
+code to analyze when the user queries the source program.
 %Locations are very important for the compiler/$merlin$ to point to the right
 %place when reporting errors/warnings.
-There are two location invariants, which PPXs need to respect:
+
+To resolve this issue, we've instituted two location invariants for the results
+of PPX transformations. These can be configured in the $ppxlib$ driver as
+automatic checks, or left to PPX authors to enforce.
 \begin{itemize}
-\item Parent-child nodes are nested wrt. locations.
+\item The locations of child nodes are contained by the locations of their parents.
 \item The locations of sibling nodes don't overlap.
 \end{itemize}
```

Comment:
Maybe add a mention on ghost locations?

```suggestion
\end{itemize}

It is also possible to escape those invariants by using "ghost locations". Using ghost locations for nodes that borrow their location to another node is a good practice, to avoid confusing Merlin.
```

Reply:
I think the last sentence might come across a bit misleading. Isn't it more like "each position should be contained in the locations of a unique sequence of parent-child nodes"?